// @aws-lambda-powertools/logger/lib/esm/formatter/LogItem.js
import merge from 'lodash.merge';
class LogItem {
    constructor(params) {
        this.attributes = {};
        // Add attributes in the log item in this order:
        // - Base attributes supported by the Powertool by default
        // - Persistent attributes provided by developer, not formatted (done later)
        // - Ephemeral attributes provided as parameters for a single log item (done later)
        this.addAttributes(params.attributes);
    }
    addAttributes(attributes) {
        merge(this.attributes, attributes);
        return this;
    }
    getAttributes() {
        return this.attributes;
    }
    prepareForPrint() {
        this.setAttributes(this.removeEmptyKeys(this.getAttributes()));
    }
    removeEmptyKeys(attributes) {
        const newAttributes = {};
        for (const key in attributes) {
            if (attributes[key] !== undefined &&
                attributes[key] !== '' &&
                attributes[key] !== null) {
                newAttributes[key] = attributes[key];
            }
        }
        return newAttributes;
    }
    setAttributes(attributes) {
        this.attributes = attributes;
    }
}
export { LogItem };
