// @aws-lambda-powertools/parameters/lib/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Transform = exports.TRANSFORM_METHOD_AUTO = exports.TRANSFORM_METHOD_BINARY = exports.TRANSFORM_METHOD_JSON = exports.DEFAULT_MAX_AGE_SECS = exports.APPCONFIG_TOKEN_EXPIRATION = void 0;
const DEFAULT_MAX_AGE_SECS = 5;
exports.DEFAULT_MAX_AGE_SECS = DEFAULT_MAX_AGE_SECS;
const TRANSFORM_METHOD_JSON = 'json';
exports.TRANSFORM_METHOD_JSON = TRANSFORM_METHOD_JSON;
const TRANSFORM_METHOD_BINARY = 'binary';
exports.TRANSFORM_METHOD_BINARY = TRANSFORM_METHOD_BINARY;
const TRANSFORM_METHOD_AUTO = 'auto';
exports.TRANSFORM_METHOD_AUTO = TRANSFORM_METHOD_AUTO;
const APPCONFIG_TOKEN_EXPIRATION = 23 * 60 * 60 * 1000 + 45 * 60 * 1000; // 23 hrs 45 min
exports.APPCONFIG_TOKEN_EXPIRATION = APPCONFIG_TOKEN_EXPIRATION;
/**
 * Transform methods for values retrieved by parameter providers.
 */
const Transform = {
    /**
     * Transform the retrieved value using `JSON.parse`.
     */
    JSON: TRANSFORM_METHOD_JSON,
    /**
     * Transform a base64-encoded value from `Uint8Array` to `string`.
     */
    BINARY: TRANSFORM_METHOD_BINARY,
    /**
     * Automatically detect the transform method based on the parameter' name suffix.
     */
    AUTO: TRANSFORM_METHOD_AUTO,
};
exports.Transform = Transform;
